decompressed/base/etc/init.d/rootdevice

```diff
@@ -307,17 +307,27 @@ check_and_apply_obp() {
         md5_bank_1=$(md5sum "$bank1_mtd" | awk '{ print $1 }')
         md5_bank_2=$(md5sum "$bank2_mtd" | awk '{ print $1 }')
         if [ "$md5_bank_1" == "$md5_bank_2" ]; then
-          logger_command "Cloning complete, moving overlay..."
-          echo "  Cloning complete, moving overlay..." >/dev/console
+          logger_command "Cloning complete, moving to bank_2 overlay..."
+          echo "  Cloning complete, moving to bank_2 overlay..." >/dev/console
           # Clean any existing overlay for bank_2 firmware
           rm -rf /overlay/bank_2
+          if [ ! -f /etc/config/modgui ]; then
+            logger_command "Seems first install from bank_1 saving GUI to root and setting install flags..."
+            echo "  Seems first install from bank_1 saving GUI to root  and setting install flags..." >/dev/console
+            [ -f /tmp/GUI.tar.bz2 ] && mv /tmp/GUI.tar.bz2 /root/GUI.tar.bz2
+            [ -f /tmp/GUI_dev.tar.bz2 ] && mv /tmp/GUI.tar.bz2 /root/GUI.tar.bz2
+            touch /root/.reapply_due_to_upgrade
+            touch /root/.install_gui
```

Comment:
Non mi è chiaro il senso di creare qui questi file. Stai assumendo una prima installazione, al prossimo riavvio `.install_gui` verrà creato a suo tempo dentro `install_gui()`.

Inoltre da come è scritto adesso sembra che `install_gui()` preveda sempre di avere entrambi i file prima di svolgere qualsiasi operazione. Fai sempre un touch di `.reapply_due_to_upgrade` prima di chiamare `install_gui()` al di fuori di essa. Fai sempre un touch di `.install_gui` all'interno di essa. Fai sempre un rm di entrambi quando va tutto a buon fine. Se durante una prima installazione questi file devo esistere allora potete metterli direttamente nel tar. Se già ce li avete messi, allora non ha senso crearli qui.

Cambiando questa riga in `boot()`
`[ -z "$force" ] && force=0`
con
`{ [ -z "$force" ] && force=0 } || touch /root/.install_gui`
puoi buttare tutta il ramo else dell'if immediatamente successivo, ed anche rimuovere il touch senza senso dello stesso file fatto dentro `install_gui()`. E' assolutamente inguardabile vedere applicata una flag, normalmente usata come condizione di ingresso ad un funzione, all'interno della stessa.

Se modgui_scripts contiene un postreq che pulisce la flag `.install_gui` nel caso vada tutto a buon fine, allora è sempre lì che va pulita anche .reapply_duo_to_upgrade

Reply:
`.install_gui` serve per assicurarci che è arrivato fino in fondo al processo, deve essere inserito runtime altrimenti non ha senso

> `.install_gui` dovrebbe già esistere visto che nulla di norma lo crea prima di chiamare `install_gui()`

no, non esiste perche non ci arriva a settarla dato che, se la installi da bank1, fa il reboot prima di arrivare lì, il `.reapply_due_to_upgrade` ci serve per indicare di reestrarre il tar (il motivo è che viene persa la roba che viene estratta in /tmp su prima installazione se vieni da bank1)

Reply:
> `.install_gui` serve per assicurarci che è arrivato fino in fondo al processo, deve essere inserito runtime altrimenti non ha senso

Qualcosa non torna perché nel flusso standard senza Force nessuno la crea su prima installazione. Deve essere nel tar in partenza. 

> no, non esiste perche non ci arriva a settarla dato che, se la installi da bank1, fa il reboot prima di arrivare lì, il

Lì dove? Non esiste nessun posto fuori da install gui in cui avviene questa cosa, il codice che lo fa te e dentro la funzione (o spostato fuori come nel suggerimento di sopra) e riguarda solo l'avvio con force.

Includi quel flag nel tar,  o crealo a runtime anche su esecuzioni senza force. Come è adesso è un popocchio tremendo. Decidi formalmente se deve fungere da condizione di ingresso o da side-effect di stato della rispettiva funzione. Nel primo caso basta farlo preesistere nel tar o crearlo a runtime con force. Nel secondo caso allora nulla al di fuori della funzione deve farvi riferimento, quindi ogni istruzione di creazione o valutazione sulla sua esistenza va fatta tassativamente dentro quella funzione che ci si aspetta termini con successo rimuovendo. 

 `.reapply_due_to_upgrade` ci serve per indicare di reestrarre il tar (il motivo è che viene persa la roba che viene estratta in /tmp su prima installazione se vieni da bank1)

Allora devi mettere il sotto sieme di ciò che va estratto dopo il reboot dentro un archivio a parte incluso assieme ai file base in un secondo. Il sotto-archivio basta chiamarlo come scritto nel codice sotto il mount - mi pare sysupgrade.tgz, verra estratto al boot senza necessità di alcuna flag. Essendo parte del tar originale messo in /tmp verrebbe già estratto in /root e non avresti qui bisogno di codice aggiuntivo.

Reply:
> Qualcosa non torna perché nel flusso standard senza Force nessuno la crea su prima installazione. Deve essere nel tar in partenza.

e così deve essere, senza flag force rootdevice non fa praticamente niente, quella flag deve esser se c'è una richista di install in corso o in sospeso, al tempo stesso non vedo il problema ad utilizzarla sia come condizione di ingresso che come check di uscita


> Allora devi mettere il sotto sieme di ciò che va estratto dopo il reboot dentro un archivio a parte incluso assieme ai file base in un secondo.

Non ne avevo minimamente voglia, tanto per ora l'idea è che il pacchetto in root deve starci sempre, ho voluto semplicemente riutilizzare le funzione di install che già c'era e che ero sicuro funzionasse perchè non avevo voglio di aggiungere altro codice


Sia chiaro, sono sicuramente logiche migliorabili (come tante altre che ci sono in giro per tutto il progetto), ma fin quando non dimostrano criticità e funzionano stan bene li come sono

Reply:
È un problema di design scadente che ha ripercussioni su altre parti di codice che se non risolvi alla radice ti costringe a mettere continuamente pezze ovunque. Non è normale che per mettere una funzione (concettualmente separata da quello che stai facendo in questo punto del programma) in condizione di operare come deve tu sia costretto ad aggiungere istruzioni ad essa correlate un po' ovunque. Se non la risolvi subito ti crei una situazione e di technical debt non indifferente che più passa il tempo e più sarà costoso in termini di sforzo e tempo risolvere. Se provi a disegnare un grafo di interdipendenza tra i vari moduli di codice di accorgi di stare accoppiando la funzione che applica il planning ad una logica che pertiene esclusivamente all'installazione, è un arco del grafo che può non essere aggiunto e pertanto aggiungerlo è sempre in linea di principio un errore, e ogni volta che accade da qualche parte nel mondo un ingegnere del software muore. 

Fidati, risolvi questo problema il prima possibile finché è semplice, altrimenti dovrai ritornare a fare ritocchi ad ogni nuova modifica strutturale.

Il modo migliore per iniziare ad arginare la falla della flag è inserirla nel tarball, rimuovere il codice che la crea da dentro la funzione ibstall_gui() e rimuovere quel ramo else per intero da boot(). A quel punto il funzionamento diventa lineare, install_gui non viene mai chiamata almeno che non sia una prima installazione o una reinstalla ione con force. È esattamente quello che fate ora ma con una ventina di righe di codice confuse in meno.

Il problema di reapply.eccc lo vediamo dopo

Reply:
forse ancora non ti ho fatto capire:
Sono il primo a cui piace ottimizzare e reingegnerizzare per fare le cose per bene (e di sicuro, anche guardando il livello di codice buttato giu da tch, non è proprio il progetto dove dimostrare sta roba :D), in quel momento ho solo pensato "fanculo fa schifo sta roba ma basta che funziona per ora sta bene cosi", l'obbiettivo era arrivare a chiudere il PR poi se ne parla anche perchè c'erano altre logiche che includevano anche gli upgrade della GUI che non avevo voglia di ricontrollare

![ezgif com-optimize](https://user-images.githubusercontent.com/4574593/80809054-2e4d7a80-8bc1-11ea-8474-1dc4d2552b4a.gif)

Reply:
Ho tolto quell'if inutile e reso un po piu lineare il ragionamento almeno in rootdevice, ho visto ci sono altre cose senza senso un 01_prereq... ma faccio finta di non averle viste 🙈

secondo me possiamo iniziare a mergiare e poi se ne parla